### ISSUE TITLE
Machine Learning

### ISSUE THREAD

**Issue by rainit2006** [rainit2006/Python-room#5]

**Comment by rainit2006** [comment]
Quandl： 金融、経済などの数値データの検索エンジンで、さまざまなソースから得られたデータを検索し、グラフや表を表示させることができる。またデータはJSON、CSVなどの形式でダウンロードしたり、Plotlyなどのサービスに取り込ませることもできる。


scikit-learn： オープンソースの機械学習ライブラリで、分類や回帰、クラスタリングなどの機能が実装されています。

**Comment by rainit2006** [comment]
数据的标准化(Standardization)、平均移除法（mean removal）和方差归一化（variance scaling）

在机器学习的目标函数（objective function），比如SVM中的RBF或者线性模型中的L1和L2正则项，其中使用的元素，前提是所有的feature都是以0为中心，且方差的阶（order）都一致。如果一个feature的方差，比其它feature的阶（order）都大，那么它将在目标函数中占支配地位，从而使得estimator从其它feature上学习不到任何东西。
http://d0evi1.com/sklearn/preprocessing/

**Comment by rainit2006** [comment]
**BLAS, LAPACK**
![image](https://user-images.githubusercontent.com/12871721/29399426-2817bf4e-8364-11e7-850d-306b9bebaee4.png)


**練習用データ**
Kaggle : https://www.kaggle.com/
UCI : https://archive.ics.uci.edu/ml/datasets.html

**Comment by rainit2006** [comment]
**向量机(以下简称SVM)算法**
http://www.cnblogs.com/pinard/p/6117515.html
scikit-learn中SVM的算法库分为两类，一类是分类的算法库，包括SVC， NuSVC，和LinearSVC 3个类。
另一类是回归算法库，包括SVR， NuSVR，和LinearSVR 3个类。相关的类都包裹在sklearn.svm模块之中。

在scikit-learn中，内置的核函数一共有4种，当然如果你认为线性核函数不算核函数的话，那就只有三种。
　1）线性核函数（Linear Kernel）表达式为：K(x,z)=x∙z，就是普通的内积，LinearSVC 和 LinearSVR 只能使用它。

　　　　2)  多项式核函数（Polynomial Kernel）是线性不可分SVM常用的核函数之一，表达式为：（K(x,z)=（γx∙z+r)d ，其中，γ,r,d都需要自己调参定义,比较麻烦。

　　　　3）高斯核函数（Gaussian Kernel），在SVM中也称为径向基核函数（Radial Basis Function,RBF），它是libsvm默认的核函数，当然也是scikit-learn默认的核函数。表达式为：K(x,z)=exp(γ||x−z||2)， 其中，γ大于0，需要自己调参定义。

　　　　4）Sigmoid核函数（Sigmoid Kernel）也是线性不可分SVM常用的核函数之一，表达式为：（K(x,z)=tanh（γx∙z+r)， 其中，γ,r都需要自己调参定义。

　　　　一般情况下，对非线性数据使用默认的高斯核函数会有比较好的效果，如果你不是SVM调参高手的话，建议使用高斯核来做数据分析。

**Comment by rainit2006** [comment]
Clustering and Unsupervised machine learning
**Flat and Hierarchical** 
With Flat clustering, the scientist tells the machine how many classes/clusters to find. 
With Hierarchical clustering(层次化的聚类), the machine figures out the groups and how many. 得出来的结构是一棵树.

The field of "Big Data Analysis" is generally a prime area for clustering. 

- K-Means algorithm
The idea of K-Means is to attempt to cluster a given dataset into K clusters. 
1. Take entire dataset, and set, randomly, K number of centroids. Centroids are just the "centers" of your clusters.
2.  Calculate distance of each featureset to the centroids, and classify each featureset as the centroid class closest to it. Centroid classes are arbitrary (任意), you will likely just call the first centroid 0, the second centroid 1...and so on.
3. Once you have classified all data, now you take the "mean" of the groups, and set the new centroids as the mean of their associated groups.
4. Repeat #1, 2 and #3 until you are optimized. 

```
clf = KMeans(n_clusters=2)
clf.fit(X)

centroids = clf.cluster_centers_
labels = clf.labels_

colors = ["g.","r.","c.","y."]
for i in range(len(X)):
    plt.plot(X[i][0], X[i][1], colors[labels[i]], markersize = 10)
plt.scatter(centroids[:, 0],centroids[:, 1], marker = "x", s=150, linewidths = 5, zorder = 10)
plt.show()
```


-----------------------------------------------
Handling Non-Numerical Data for Machine Learning

**Comment by rainit2006** [comment]
■train_test_split
データセットから取り出した X y をさらに、「トレーニング用」と「テスト用」のデータに分割します。
![image](https://user-images.githubusercontent.com/12871721/43057955-08c550d4-8e80-11e8-8b1b-c646cb0f9072.png)
```
from sklearn.model_selection import train_test_split

(X_train, X_test, y_train, y_test) = train_test_split( X, y, test_size=0.3, random_state=0, )
```
train_test_split には以下のような引数を与えます。
•第一引数: 特徴行列 X
•第二引数: 目的変数 y
•test_size=: テスト用のデータを何割の大きさにするか◦test_size=0.3 で、3割をテスト用のデータとして置いておけます
•random_state=: データを分割する際の乱数のシード値◦同じ結果が返るように 0 を指定していますが、これは勉強用であり普段は指定しません

**Comment by rainit2006** [comment]
为了提高精度，数据的预处理（Preprocessing）和 Feature Engineering(特征工程) 很重要！
In Data Preprocessing, we usully do something like:
•deal with outlier
•categorical variable encoding.(One-Hot Encoding)
•text data embedding.

IN Feature Engineering, we usully do something like:
•Feature Extraction(Create some new feature)
•Feature Selection(by feature correlation)
